### PR TITLE
JAMES-2665 Write a first basic integration test for the vault

### PR DESCRIPTION
**Pull request by chibenwa**
Note that vault + FileMailRepository combination was broken:

`AttributeValue<Collection<AttributeValue<?>>>` is not serializable. To work around this issue and not completlty break retro-comatibility, I chose to leverage java serialization for AttributeValue usiung the available JSON serialization, which solves the issue.

Other issues spotted and solved along the way.

### CODE REVIEW

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
 import java.util.Iterator;
```
Comment (mbaechler) [inline server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: Why do you think so? MailImpl contains a resource `message` that we should probably don't try to evaluate in equals/hashcode. So I would argue the equality should not be provided by MailImpl.

File: server/container/core/src/test/java/org/apache/james/server/core/MailImplTest.java [hunk 2]
```diff
@@ -292,4 +301,68 @@ void mailImplShouldNotAllowSettingEmptyName() {
...
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
+        objectOutputStream.writeObject(mail);
+        objectOutputStream.close();
+        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
+        ObjectInputStream objectInputStream = new ObjectInputStream(byteArrayInputStream);
+        Object object = objectInputStream.readObject();
+
+        assertThat(object)
+            .isInstanceOf(MailImpl.class)
+            .isEqualTo(mail);
```
Comment (mbaechler) [inline server/container/core/src/test/java/org/apache/james/server/core/MailImplTest.java]: if you introduced equality for that purpose, you can do a compare field by field ignoring `message` field instead.

File: server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java (hunk 1 above)
Comment (chibenwa) [reply server/container/core/src/main/java/org/apache/james/server/core/MailImpl.java]: It was just convenient to implement the "serializable" concern.

Agree, let's do a "field by field" equality

File: server/container/core/src/test/java/org/apache/james/server/core/MailImplTest.java (hunk 2 above)
Comment (chibenwa) [reply server/container/core/src/test/java/org/apache/james/server/core/MailImplTest.java]: Yes I did ^^

**Comment by chibenwa** [comment]
```

[7fc7c09cc53244b996f6fea8f4e238096b3f88a1] [ERROR]   SwiftKeystone2ObjectStorageBlobsDAOBuilderTest.builtBlobsDAOCanStoreAndRetrieve:92 » NoSuchElement
[7fc7c09cc53244b996f6fea8f4e238096b3f88a1] [ERROR]   SwiftKeystone3ObjectStorageBlobsDAOBuilderTest.builtBlobsDAOCanStoreAndRetrieve:128 » NoSuchElement
[7fc7c09cc53244b996f6fea8f4e238096b3f88a1] [ERROR]   SwiftKeystone3ObjectStorageBlobsDAOBuilderTest.builtBlobsDAOCanStoreAndRetrieve:128 » NoSuchElement
[7fc7c09cc53244b996f6fea8f4e238096b3f88a1] [ERROR]   SwiftKeystone3ObjectStorageBlobsDAOBuilderTest.builtBlobsDAOCanStoreAndRetrieve:128 » NoSuchElement
```

test this please

**Comment by chibenwa** [comment]
:green_apple:  before ci restart

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
Read it... Feels like I saw some of those commits not marked `NO_REVIEW` in other PRs as well. So it's hard to evaluate it. But it will probably make more sense I guess when some other PRs will be merged and this one rebased?

**Comment by chibenwa** [comment]
`it will probably make more sense I guess when some other PRs will be merged and this one rebased?`

Yes exactly

**Comment by chibenwa** [comment]
test this please

**Comment by openpaas-james** [comment]
Build finished. No test results found.

**Comment by chibenwa** [comment]
:green_apple:

**Comment by Arsnael** [comment]
test this please

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 3]
```diff
@@ -40,10 +42,13 @@
 
 /** 
  * Strong typing for attribute value, which represents the value of an attribute stored in a mail.
+ *
+ * Note that this class is Java serializable for compatibility purposes. Use at your own risks and always prefer relying on the JSON representation.
  * 
  * @since Mailet API v3.2
  */
-public class AttributeValue<T> {
+@Immutable
```
Comment (mbaechler) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]:
```
/**
 * The class to which this annotation is applied is immutable. This means that
 * its state cannot be seen to change by callers. Of necessity this means that
 * all public fields are final, and that all public final reference fields refer
 * to other immutable objects, and that methods do not publish references to any
 * internal state which is mutable by implementation even if not by design.
 * Immutable objects may still have internal mutable state for purposes of
 * performance optimization; some state variables may be lazily computed, so
 * long as they are computed from immutable state and that callers cannot tell
 * the difference.
 * <p>
 * Immutable objects are inherently thread-safe; they may be passed between
 * threads or published without synchronization.
 */
```

So it's not applicable

File: mailet/api/src/main/java/org/apache/mailet/AttributeValue.java [hunk 4]
```diff
@@ -40,10 +42,13 @@
...
+public class AttributeValue<T> implements Serializable {
```
Comment (mbaechler) [inline mailet/api/src/main/java/org/apache/mailet/AttributeValue.java]: There's 28 pages to read from Effective Java before you consider seriously this solution. There's so many pitfalls that I can't recall them all.

**Comment by chibenwa** [comment]
So let's propose two alternatives:

 - Not supporting the vault feature on top of `FileMailRepository`, which we should eventually deprecate....
 - [1] Changing MailImpl serialization:
    - Try to deserialize attibutes as JSON and fallback to previous code overwize.
 - Over solutions are welcome

I propose : 
 - Let's not support [1] for now and split this PR

@mbaechler any thoughts?

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**